client.Enqueue - prevent empty task typename

Luqqk opened the pull request:
Hi @hibiken! Thanks for this library and sharing your work 👍 

When I was trying to get familiar with the code base and test out some of the features locally I also decided to contribute and help a little bit. I picked #297 as something that appears easy to start with. Please let me know if I missed something out or the solution should be improved/adjusted. Keep in mind that currently you're still able to queue and consume tasks with typenames like `" "` (just a single whitespace). This can be a feature but maybe also should be handled differently?

Thanks again and I'm looking forward to contribute in future!

crossworth commented:
Hello @Luqqk , great work!

If you don't mind I would like to make a few comments.

> Keep in mind that currently you're still able to queue and consume tasks with typenames like " " (just a single whitespace). This can be a feature but maybe also should be handled differently?

Currently on the handler registration the check is with ` == ""` as well, so maybe this is not a problem, but maybe we should use `TrimSpace` in the future, let's see what @hibiken thinks about it.
https://github.com/hibiken/asynq/blob/73f930313cef4434c7a62f54bece8051788fa89c/servemux.go#L101

Maybe the error name should be:  `ErrEmptyTypename` just to make it more precise.

I have one more consideration, maybe the error should be handled on the `asynq/internal/rdb/rdb.go` as well, since `ErrDuplicateTask` is handled in a similar way, but I dont see why right now, since rdb is internal and the only client is `client.go`.

Luqqk commented:
Thanks for quick feedback @crossworth!

I'll adjust the error name as you suggested. 

> I have one more consideration, maybe the error should be handled on the asynq/internal/rdb/rdb.go as well, since ErrDuplicateTask is handled in a similar way, but I dont see why right now, since rdb is internal and the only client is client.go.

Yup, I'm not sure about this one as well and that's the reason I added it only in `client.go` for now.

Let's wait for @hibiken, maybe he'll shed some light on it.

hibiken commented on client.go:
```diff
@@ -178,6 +178,9 @@ func (d processInOption) Value() interface{} { return time.Duration(d) }
 var ErrDuplicateTask = errors.New("task already exists")
 
+// ErrEmptyTypeTask indicates that task's typename is not specified.
+var ErrEmptyTypeTask = errors.New("task typename not specified")
```
I don't expect many users encountering this error, so surfacing a sentinel error value via public API seems like an overkill. Let's not add this to package API yet. We can always add this to public API if needed :)

Also, another nit-pick: can we change the error message to `"task typename cannot be empty"` (reason: Technically user has *specified* typename, it's just empty)

Luqqk replied:
Thanks for the feedback. I fully agree that the error message you suggested sounds better. Adjusted it as requested.

hibiken commented on client.go:
```diff
@@ -266,6 +269,9 @@ func (c *Client) Close() error {
 //
 // If no ProcessAt or ProcessIn options are provided, the task will be pending immediately.
 func (c *Client) Enqueue(task *Task, opts ...Option) (*TaskInfo, error) {
+	if task.Type() == "" {
```
Like @crossworth suggested, let's `TrimSpace` here so that we can check for any blank strings.
Please make sure to update corresponding check in `servemux.go` as well!

Luqqk replied:
Added `TrimSpace` in `client.go` and `servemux.go`

hibiken reviewed (changes requested):
Thank you for the contribution @Luqqk!

@crossworth thanks for reviewing it so quickly, I really appreciate your comment as always :)

Let's defer exposing sentinel error value for this case, I expect very few users will encounter this error anyways.

hibiken commented on client.go:
```diff
@@ -266,6 +270,9 @@ func (c *Client) Close() error {
 // If no ProcessAt or ProcessIn options are provided, the task will be pending immediately.
 func (c *Client) Enqueue(task *Task, opts ...Option) (*TaskInfo, error) {
+	if strings.TrimSpace(task.Type()) == "" {
+		return nil, fmt.Errorf("%w", ErrEmptyTypename)
```
this should be sufficient, since we don't want to export a sentinel error value for this case.
```go
return nil, fmt.Errorf("task typename cannot be empty")
```

Luqqk replied:
Ok, now I see what you meant. Thanks for clarifying.

hibiken commented on client.go:
```diff
@@ -178,6 +179,9 @@ func (d processInOption) Value() interface{} { return time.Duration(d) }
 var ErrDuplicateTask = errors.New("task already exists")
 
+// ErrEmptyTypename indicates that task's typename is empty.
+var ErrEmptyTypename = errors.New("task typename cannot be empty")
```
Like I said in the last comment, I don't think we should export this error value from package API (I don't expect many users encountering this error anyway).

hibiken commented on client_test.go:
```diff
@@ -585,6 +585,11 @@ func TestClientEnqueueError(t *testing.T) {
+			desc: "With empty task typename",
+			task: NewTask("", h.JSON(map[string]interface{}{})),
+			opts: []Option{},
+		},
```
another case for blank line?

Luqqk replied:
Added another test case.

hibiken reviewed (changes requested):
Thanks for fixing so quickly. I left a few comments, please address those as well.
Also, I forgot in my last review, but please add an entry in CHANGELOG.md file.

Luqqk commented:
I adjusted the code as requested. Also added the changelog entry. Small suggestion - would be nice to add an info in the "contribution guide" that changelog entry should be a part of the PR ;)
